domain/application/leadershihp_test.go

```diff
@@ -0,0 +1,206 @@
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/lease"
+	"github.com/juju/juju/core/status"
+	corestorage "github.com/juju/juju/core/storage"
+	jujuversion "github.com/juju/juju/core/version"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/application/charm"
+	"github.com/juju/juju/domain/application/service"
+	"github.com/juju/juju/domain/application/state"
+	"github.com/juju/juju/domain/secretbackend/errors"
+	changestreamtesting "github.com/juju/juju/internal/changestream/testing"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+	"github.com/juju/juju/internal/storage"
+	"github.com/juju/juju/internal/storage/provider"
+	coretesting "github.com/juju/juju/internal/testing"
+	"github.com/juju/juju/internal/uuid"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+)
+
+type leadershipSuite struct {
+	changestreamtesting.ModelSuite
+
+	leadership *MockChecker
+}
+
+var _ = gc.Suite(&leadershipSuite{})
+
+func (s *leadershipSuite) SetUpTest(c *gc.C) {
+	s.ModelSuite.SetUpTest(c)
+
+	modelUUID := uuid.MustNewUUID()
+	err := s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		_, err := tx.ExecContext(ctx, `
+			INSERT INTO model (uuid, controller_uuid, target_agent_version, name, type, cloud, cloud_type)
+			VALUES (?, ?, ?, "test", "iaas", "test-model", "ec2")
+		`, modelUUID.String(), coretesting.ControllerTag.Id(), jujuversion.Current.String())
+		return err
+	})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *leadershipSuite) TestSetApplicationStatusForUnitLeader(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	done := make(chan struct{})
+	s.leadership.EXPECT().WaitUntilExpired(gomock.Any(), "foo", gomock.Any()).DoAndReturn(func(ctx context.Context, s string, c chan<- struct{}) error {
+		close(c)
+		// Block until the call is done.
+		<-done
+		return nil
+	})
+	s.leadership.EXPECT().Token("foo", "foo/666").Return(leaseToken{})
+
+	svc := s.setupService(c, nil)
+
+	u1 := service.AddUnitArg{
+		UnitName: "foo/666",
+	}
+	appID := s.createApplication(c, svc, "foo", u1)
+
+	err := svc.SetApplicationStatusForUnitLeader(context.Background(), "foo/666", &status.StatusInfo{
+		Status: status.Active,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	close(done)
+
+	appStatus, err := svc.GetApplicationStatus(context.Background(), appID)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(appStatus.Status, gc.Equals, status.Active)
+}
+
+func (s *leadershipSuite) TestSetApplicationStatusForUnitLeaderNotTheLeader(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	done := make(chan struct{})
+	s.leadership.EXPECT().WaitUntilExpired(gomock.Any(), "foo", gomock.Any()).DoAndReturn(func(ctx context.Context, s string, c chan<- struct{}) error {
+		close(c)
+		// Block until the call is done.
+		<-done
+		return nil
+	})
+	s.leadership.EXPECT().Token("foo", "foo/666").Return(leaseToken{
+		error: errors.NotValid,
+	})
```

Comment:
It would be useful imo to include an integration test that integrates with a real leadership checker. Especially given it's security-related.

How feasible would this be?

Reply:
That said, you have a lot of tests in this PR anyway so this is completely optional. Especially if this turns out to be a big ask.

Reply:
I 100% agree, it's a lot of work and that's coming from me saying that.